pkg/agent/secondarynetwork/podwatch/k8s/podwatch_controller.go

```diff
@@ -0,0 +1,318 @@
+	netdefutils "github.com/k8snetworkplumbingwg/network-attachment-definition-client/pkg/utils"
+
+	cnipodcache "antrea.io/antrea/pkg/agent/secondarynetwork/cnipodcache"
+	ipam "antrea.io/antrea/pkg/agent/secondarynetwork/ipam"
+	cniserver "antrea.io/antrea/pkg/agent/cniserver"
+)
+
+const (
+	controllerName = "AntreaAgentPodWatchController"
+	minRetryDelay  = 2 * time.Second
+	maxRetryDelay  = 120 * time.Second
+	numWorkers     = 4
+)
+
+const (
+	networkAttachDefAnnotationKey = "k8s.v1.cni.cncf.io/networks"
+	sriov                         = "sriov"
+	netDev                        = "eth1"
+)
+
+type PodWatchController struct {
+	kubeClient         clientset.Interface
+	netAttachDefClient netdefclient.K8sCniCncfIoV1Interface
+	queue              workqueue.RateLimitingInterface
+	podInformer        cache.SharedIndexInformer
+	podLister          corelisters.PodLister
+	nodeName           string
+	podCache           cnipodcache.CNIPodInfoStore
+	cniServer          *cniserver.CNIServer
+}
+
+func NewPodWatchController(kubeClient clientset.Interface,
+	netAttachDefClient netdefclient.K8sCniCncfIoV1Interface,
+	podInformer cache.SharedIndexInformer,
+	resyncPeriod time.Duration,
+	nodeName string,
+	podCache cnipodcache.CNIPodInfoStore,
+	cniServer *cniserver.CNIServer) *PodWatchController {
+
+	pwc := PodWatchController{
+		kubeClient:         kubeClient,
+		netAttachDefClient: netAttachDefClient,
+		podInformer:        podInformer,
+		podLister:          corelisters.NewPodLister(podInformer.GetIndexer()),
+		nodeName:           nodeName,
+		podCache:           podCache,
+		cniServer:          cniServer,
+	}
+
+	podInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    pwc.enqueuePod,
+			DeleteFunc: pwc.enqueuePod,
+			UpdateFunc: func(old, cur interface{}) { pwc.enqueuePod(cur) },
+		},
+		resyncPeriod,
+	)
+
+	pwc.queue = workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "podwatchcontroller")
+	return &pwc
+}
+
+func podKeyGet(pod *corev1.Pod) string {
+	return pod.Namespace + "/" + pod.Name
+}
+
+func (podWatcher *PodWatchController) enqueuePod(obj interface{}) {
+	var err error
+	pod, isPod := obj.(*corev1.Pod)
+	if !isPod {
+		podDeletedState, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			klog.ErrorS(err, "Unexpected object received:", obj)
+			return
+		}
+		pod, ok := podDeletedState.Obj.(*corev1.Pod)
+		if !ok {
+			klog.ErrorS(err, "DeletedFinalStateUnknown object is not of type Pod: ", podDeletedState.Obj, pod)
+			return
+		}
+	}
+	podKey := podKeyGet(pod)
+	podWatcher.queue.Add(podKey)
+}
+
+// handleAddUpdatePod handles Pod Add, Update events and updates annotation if required.
+func (podWatcher *PodWatchController) handleAddUpdatePod(obj interface{}) error {
+	var ret bool
+	pod := obj.(*corev1.Pod)
+
+	secondaryNetwork, ret := checkForPodSecondaryNetworkAttachement(pod)
```

Comment:
In case, you missed the previous comments:

> You can check the old version of the Pod obj to know if the CNF annotation got removed from the Pod.
> Check: https://github.com/antrea-io/antrea/blob/main/pkg/agent/nodeportlocal/k8s/npl_controller.go#L211